### ISSUE TITLE
Tabbed Widget

### ISSUE THREAD

**Issue by VisenDev** [david-vanderson/dvui#130]
Tabs seem like a fairly common ui construct. It might be worth adding a widget to support this in dvui.}

Proposed api

```zig
{
  const TabNames = enum{home, edit, help};
  const tabs = try dvui.tabbed(@src(), TabNames, TabbedOptions(TabNames){
    .tab_render_fn = &default_tab_render_fn,
    .location = .top,
   }, .{});
  defer tabs.deinit();

  switch(tabs.getActiveTab()) {
     .home => {
        //render home menu
     },
     .edit => {
       //render edit menu
     },
     .help => {
        //render help menu
     }
  }
}
```

The usage of a `tab_render_fn` lets tabs support more complex widgets like icons - rather than being limited to just text.

**Comment by david-vanderson** [comment]
Good idea. I remember @JosephABudd made a tabbed interface.  Is that still available?  I think it used MenuItems as the tabs, and looked something like this (ignore the red borders):
https://github.com/david-vanderson/dvui/discussions/67

Instead of a callback, I'd rather the api be that the code renders each tab directly.  There's not that much functionality, but likely a lot of visual details to get right.  Which corners of the tabs to round, how to show the active one differently, what happens if there are too many to fit in a single row/col.

**Comment by JosephABudd** [comment]
The widget is in the widget/zip folder.
[widget.zip](https://github.com/user-attachments/files/17046517/widget.zip)

Below is the tab-bar being initeilaized:
```zig
        self.tabs = try Tabs.init(startup, self_as_container, tabs_options);
        errdefer self.deinit();

        // Create 1 of each type of tab.

        try self.AddNewAddTab(true);
```

Blow is the AddNewAddTab fn.
```zig
    pub fn AddNewAddTab(
        self: *Screen,
        selected: bool,
    ) !void {
        // The Add tab uses this screen's Add panel for content.
        const panel: *AddPanel = try AddPanel.init(
            self.allocator,
            self.window,
            self.main_view,
            self.messenger.?,
            self.exit,
            self.state.?.*,
        );
        const panel_as_content: *Content = try panel.asContent();
        errdefer panel.deinit();
        const tab: *Tab = try Tab.init(
            self.tabs.?,
            self.main_view,
            panel_as_content,
            .{
                // KICKZIG TODO:
                // You can override the options for the Add tab.
                //.closable = true,
                //.movable = true,
            },
        );
        errdefer {
            panel_as_content.deinit();
        }
        try self.tabs.?.appendTab(tab, selected);
        errdefer {
            self.allocator.destroy(tab);
            panel_as_content.deinit();
        }
    }
```

**Comment by JosephABudd** [comment]
The screen shots are at https://github.com/JosephABudd/kickzig

**Comment by david-vanderson** [comment]
Thanks - that helps a lot!

**Comment by JosephABudd** [comment]
I've been refactoring pretty much everything in kickzig. Here is my final working version of the widget folder. I worked on it today and fixed a couple of bigs that were in the version I uploaded previously. The edits were in api.zig and TabBarWidget.zig. 
[tabbar.zip](https://github.com/user-attachments/files/17051437/tabbar.zip)

**Comment by david-vanderson** [comment]
I've pushed an initial implementation of tabs.  See the demo window "Menus / Tabs".

How does this look?

**Comment by JosephABudd** [comment]
The demo window is very impressive. What a nice improvement.

Re the tabs:
* Flex is at times a little jumpy.
* When I use the horizontal tabs with the demo window narrowed so the horizontal menu is too small I don't get a scroll. (I'm guessing it's because the scrolling widget does not bubble events.)

I noticed the scrolling issue before and here is the solution I have been working on. I'm using a different design built from your's. Here is a simple summary. Most of my work is done but #3 below is where I have hit a wall.
 
1. Allow the FloatingWidget to bubble events to the menu so that the user can use the mouse and keyboard to scroll the menu. (I have not started this yet.)
2. Instead of moving around menu items to make room for the tab-drop, highlight the side there the tab will be dropped, like vscode.
3. Instead of dragging the selected tab, drag it's drag-widget. (This is where I've hit a wall because I can't figure out how to create a widget for the FloatingWidget to drag.)
4. Instead of dropping the selected tab, drop it's drop-widget (which for a tab, would be the tab's self).

So I have a question regarding #3 above: Do you have a quick suggestion how to create a widget that the FloatingWidget will drag?

**Comment by david-vanderson** [comment]
Thanks for taking a look!  I'm not following you totally, so help me step by step:
* Flex is not in the "Menus / Tabs" - do you mean in the reorderable lists section?  I agree it is a bit jumpy, I will look into it.
* In the "Menus / Tabs" section with the tabs you should be able to scroll the tabs (either with mouse wheel or arrow keys) - is that not working for you?

Let's get that figured out first, and then dig into the dragging stuff.

**Comment by JosephABudd** [comment]
OK, I like your tabs. Very nice.

**Comment by david-vanderson** [comment]
Good - thanks!

Can you walk me through what the scrolling issue is?  Is this a problem scrolling the reorderable lists?

**Comment by JosephABudd** [comment]
The scrolling issue was only with your Reorderable Lists.

In the screen shot see the Reorderable Horizontal Menu. I've shortened the width of that popup window so that the horizontal menu width is less than the width of it's all of it's menu items. (Your multi direction cursor is really there but it didn't show up in the screen shot. That's not the issue.) The real issue is that I can't ever scroll the menu.

So for instance, if I were to select the **zero** or **one** menu-item I could not move it between the **four** and **five** menu-items because there is no scrolling.

![dvui_horizontal_menu_narrow](https://github.com/user-attachments/assets/d97ae6a3-ced3-4992-bc25-1035013850df)

**Comment by david-vanderson** [comment]
Thanks for the screenshot and explanation!  Let me see if I can add scrolling to that.

**Comment by david-vanderson** [comment]
I've pushed a commit to try and address both the jumpiness and the horizontal scrolling.  Does that look better for you?

The horizontal scrolling was pretty easy - I added a scrollArea wrapping the reorder.

The jumpiness is more difficult.  The general advice here is "make your reorderable entries the same size in the layout direction", which is what I did.

I'm not sure there's a way to truly fix the jumpiness within the flexibility we have.  Reorderables currently work within any layout widget - they only need a sequence.  If one is being dragged, for each reorderable we ask "is the dragged one above the rectangle we just got?" and if so we ask for a second rectangle (which might not be contiguous with the first).  This results in duplicating the size of the entry being dragged over.

Fixing that would require a much more complicated layout algorithm.  Something where a widget would ask for a rectangle but then have the option of not taking it, or something like that.  If we run into more problems like this, we can revisit this option.
